qiskit_experiments/calibration_management/calibration_utils.py

```diff
@@ -12,10 +12,55 @@
 
 """Calibration helper functions"""
 
-from typing import List, Set
+from typing import List, Optional, Set
 from qiskit.pulse import ScheduleBlock, Call
 
 
+def compare_schedule_blocks(schedule1: ScheduleBlock, schedule2: ScheduleBlock) -> bool:
+    """Recursively compare schedule blocks in a parameter value insensitive fashion.
+
+    This is needed because the alignment contexts of the pulse builder creates
+    ScheduleBlock instances with :code:`f"block{itertools.count()}"` names making
```

Comment:
I wonder if the `itertools.count()` naming issue should be address in terra. @nkanazawa1989 if all the block content is equal except for the block names being different should two `ScheduleBlock`s evaluate as equal?

Even if the naming issue were addressed, this `compare_schedule_blocks` function would still be needed for the parameter equivalence we want (based on name rather than id). I wonder though if we could take all the parameters from `schedule1`, create a `params` table matching `param_a` from `schedule2` to `param_a` from `schedule1`, and the use `schedule1 == schedule2.assign_parameters(params, inplace=False)`. The benefit to this is we could use `ScheduleBlock.__eq__()` instead of relying on string equality.

Reply:
Yes, the name doesn't matter. It eventually compares DAG representation of two schedule block. So even list ordering doesn't matter.
https://github.com/Qiskit/qiskit-terra/blob/688cf6abe4ec0a2f843a63135cc6c3e9a497b2c3/qiskit/pulse/schedule.py#L1267-L1289

Reply:
I'd be happy to remove this function once we have something in Terra that we can use.

Reply:
But can't we do this now for a more thorough equality check making use of `ScheduleBlock.__eq__()`?

```python
names1 = set(p.name for p in schedule1.parameters)
names2 = set(p.name for p in schedule2.parameters)

if names1 != names2:
    return False
# guard against doubled parameters (is this necessary?)
if len(names1) != len(schedule1.parameters):
    return False
if len(names2) != len(schedule2.parameters):
    return False
    
sched2_reassigned = schedule2.assign({p.name: p for p in schedule1.parameters}, inplace=False)

return schedule1 == sched2_reassigned
```

Reply:
Hmmm I tried that out but it seems to be unable to properly compare two ECR schedules. It fails when doing `__eq__` between the two schedules on the comparison of the channels step. I suspect that the channels are somehow not being assigned properly which could be due to `CalledScheduleByName`. Have you tried that tests out with the code you suggest?

Reply:
Hmm, no, I have not tried `CalledScheduleByName`. I have not reviewed all the new updates to the PR yet. I can look at it. This is the test I was looking at:

```python
from qiskit import pulse
from qiskit.circuit import Parameter


schedules = []
for _ in range(2):
    duration = Parameter("d")
    amp = Parameter("amp")
    beta = Parameter("beta")
    sigma = Parameter("sigma")

    with pulse.build(name="xp") as xp:
        pulse.play(pulse.Drag(duration, amp, sigma, beta), pulse.DriveChannel(0))
        pulse.play(pulse.Drag(duration, amp, sigma, beta), pulse.DriveChannel(0))
    schedules.append(xp)
    
sched1, sched2 = schedules

def reassign(schedule1, schedule2):
    params1 = {p.name: p for p in schedule1.parameters}
    params2 = {p.name: p for p in schedule2.parameters}

    if set(params1) != set(params2):
        return False
    # guard against doubled parameters (is this necessary?)
    if len(params1) != len(schedule1.parameters):
        return False
    if len(params2) != len(schedule2.parameters):
        return False
        
    sched2_reassigned = schedule2.assign_parameters(
        {param2: params1[name] for name, param2 in params2.items()},
        inplace=False,
    )
    return sched2_reassigned

print("sched1 == sched2", sched1 == sched2)
print("sched1 is sched2", sched1 is sched2)
print("sched1 == reassign(sched1, sched2)", sched1 == reassign(sched1, sched2))
```

Reply:
Did you try substituting my code and running the calibration tests, or something else? Looking quickly at `CalledScheduleByName`, it might need to handle channels a bit more carefully -- allowing for the channels to have parameters and be reassigned by the `assign_parameters` method of `CalledScheduleByName`/`ScheduleBlock`, but that might not be the specific issue you were running into.

Reply:
Sounds like `CalledScheduleByName` should implement proper `__eq__`. I agree with Will that string equality is not recommended because pretty print doesn't guarantee two objects are identical. 

> Furthermore, for calibrations only the name of the parameters (as opposed to the instance) is relevant.

I'm also bit skeptical of this statement. It this is true, x and y pulse schedule should be equivalent. I'm not clear about definition of equivalence. Because `ScheduleKey` only takes name and associated qubits, this indicates we should be able to uniquely identify the schedule with these information, otherwise we may have some design error. I understand schedule here is not aware of qubits, so likely we can evaluate equality based only on the name of the most outer block, given names are unique.